### ISSUE TITLE
CFP: WireGuard node-to-node strict mode

### ISSUE THREAD

**Issue by 3u13r** [cilium/cilium#28026]
## Cilium Feature Proposal

**Is your proposed feature related to a problem?**

Node-to-node encryption is only eventual encrypted due to Endpoint/CiliumNode to IPCache propagation being delayed and only eventual consistent.

**Describe the feature you'd like**

I need a way to enforce encryption for all cluster-to-cluster (i.e. pod-to-X, node-to-X) traffic. 

**(Optional) Describe your proposed solution**

A CIDR based strict mode (similar to the pod-to-pod one).

This solution comes with some caveats:
1. The underlying threat model only makes sense if node-to-node traffic to control-plane nodes is also encrypted
2. If all node-to-node traffic must be encrypted this would include the etcd traffic (for a stacked K8s architecture). Since during `kubeadm join` the etcd connects to the other etcds one needs to allowlist its ports (2379 port and 2380) for the node CIDR.
3. Since in our architecture the kubeAPI server is only accessed via a LB with an IP outside of the nodeCIDR, there is no need to explicitly allow this connection. But different architectures have different requirements here.

I already have a working version (for us, on GCP(direct routing), AWS(vxlan), Azure(vxlan)) including tests in clilium-cli (at the moment only for direct routing, but I'll implement tunneling soon). I'll open a draft PR for you to have a look. 

Given the caveats and the current implementation I don't really see that the implementation is upstreamable. But I'm happy to do so if you think otherwise or have improvement suggestions for the implementation.

**Comment by 3u13r** [comment]
The question is what to do about those architectures which have the control-planes excluded from node-to-node encryption. One can either add some complexity to the whole thing to allow a more insecure mode in which the control-plane node traffic does not need to be encrypted or exclude those use-cases. 
I personally don't see a huge (security) benefit for the first case and therefore would opt for excluding the use-case.

In any case, I think that a node-to-node strict mode which also allows the 2nd use-case builds on top of the first. So we'd need to build the first architecture anyway.

**Comment by gandro** [comment]
> EDIT: What I think is a better solution than to disable node-to-node encryption completely is to allowlist certain ports/connections. This is what I do for etcd at least.

Yeah, I think that would be the ideal solution, it's something that has been requested quite often even outside of node-to-node encryption. However, it is probably not that easy to implement.

---

A "cheaper" solution might be to allow strict mode to allow unencrypted traffic for certain IPCache entries. For example, if an IPCache entry _is present_ and does have encryptKey=0, we could allow that traffic to not be encrypted, even if it's IP address is part of the "strict" CIDR. That approach however is problematic if IPs can be reused, i.e. if an IP of a previously unencrypted node is re-assigned to a node (or pod) that now requires strict encryption. 

I think the "allow certain ports/connections" solutions might suffer from similar problems though, depending on how dynamic or static the allowlist is.

**Comment by 3u13r** [comment]
> A "cheaper" solution might be to allow strict mode to allow unencrypted traffic for certain IPCache entries. 

I also thought about that because it is especially neat to re-use the `allow-remote-node-identities` from the pod-to-pod strict mode. Sadly, I don't think it is sufficient in e.g. the etcd's case. When the etcd talks with the other etcds there is no K8s Node entry or similar. So I don't see what we could use to differentiate the joining node's traffic from unwanted unencrypted traffic. 

Of course, using the IPCache should fix the case when the control-plane node has already joined.

**Comment by sepich** [comment]
> I need a way to enforce encryption for all cluster-to-cluster (i.e. pod-to-X, node-to-X) traffic.

Sorry for bothering, just willing to provide some userstory for this feature:

We want to create k8s cluster on Hetzner hw nodes. All nodes there have public IP, and optional internal IP. It could be only 100 internal IPs connected to one "vSwitch" and we want to run 200+ nodes. Also vSwitch connection has lower MTU, and worse latency/bandwidth between 2 same nodes, compared to measuring between their public IPs. Even if we create wireguard tunnel between public IP, tests show better results in tunnel than over "vSwitch".

So, we want to setup k8s network on public IPs, obviously everyone would want encryption in this case. And cilium has wireguard support - great! 

Starting simple - 3 nodes, with etcd/controlplane stacked via kubeadm. K8s API VIP is on internal vSwitch IP. Deploying cilium v1.14.2 in `kubeProxyReplacement: true, tunnel: vxlan` mode and `cilium-test` all passing. Just adding:
```yaml
encryption:
  enabled: true
  type: wireguard
  nodeEncryption: true
extraConfig:
  node-encryption-opt-out-labels: ""
```
And already some problems:
- wireguard uses `internal` interface for tunnels, but we want it via `external`. Figured out that it is possible to change that by editing `ciliumNode.spec.addresses.InternalIP` to external IP of each CiliumNode object. This breaks couple of tests from `cilium-test`.
- There are per-pod tunnels:
```
root@k8s-master-1:~# wg show
interface: cilium_wg0
  public key: xxx
  private key: (hidden)
  listening port: 51871
  fwmark: 0x1e00

peer: xxx
  endpoint: 10.61.16.10:55197
  allowed ips: 10.89.0.70/32, 10.61.16.10/32, 10.89.0.44/32, 10.89.0.9/32, 10.89.0.42/32, 10.89.0.89/32, 10.89.0.36/32, 10.89.0.56/32, 10.89.0.12/32
  latest handshake: 1 minute, 16 seconds ago
  transfer: 168.07 KiB received, 152.25 KiB sent

peer: xxx
  endpoint: 10.61.16.12:39104
  allowed ips: 10.89.0.192/32, 10.89.0.138/32, 10.89.0.193/32, 10.89.0.174/32, 10.89.0.172/32, 10.89.0.238/32, 10.89.0.204/32, 10.61.16.12/32, 10.89.0.167/32
  latest handshake: 1 minute, 27 seconds ago
  transfer: 59.33 KiB received, 224.46 KiB sent
```
And this cluster is empty yet. I don't believe this is scalable solution for 200+ node cluster, if each node would have specific tunnel with per-pod granularity and constantly update them keeping up with pod churn.
- Some traffic still [not enrypted](https://docs.cilium.io/en/latest/security/network/encryption/#known-issues-and-workarounds). Found that it [has been improved in v1.15](https://github.com/cilium/cilium/pull/21856) and tested on `1.15.0-pre.0` with:
```yaml
encryption:
  enabled: true
  type: wireguard
  nodeEncryption: true
  strictMode: # v1.15
    enabled: true # https://docs.cilium.io/en/latest/security/network/encryption/
    cidr: 10.89.0.0/16
extraConfig:
  node-encryption-opt-out-labels: ""
```
and it breaks in ~10min as per-docs, probably due to unable to renew encryption key from master

Unfortunately cilium implementation of wireguard is overly complicated. I believe good example of how this could been done is https://kilo.squat.ai
Workaround for case above looks like this:
- Each node runs Kilo daemonset, which configures full mesh node-to-node wireguard on public IPs
- This ends up as network interface with internal subnet shared by all nodes. (We remove 100 nodes limit of vSwitch here)
- We use this interface to run cilium on top in `tunnel: disabled` mode
- Kilo reads cilium `cluster-pool` IPAM subnet from each node, and only creates routes for node podCIDR, not for each pod. This is pretty static info, no need to constantly keep it up with pod churn
- No double encapsulation, and nodes could be in different subnets
- Cilium host firewall still works

This way all pod-to-X, node-to-X traffic uses Kilo interfaces and is encrypted. There is non-wg traffic from kubelets to k8s API VIP, and etcd traffic but it is TLS and on internal subnet. 

> A CIDR based strict mode (similar to the pod-to-pod one).

Sorry for taking your time with this long read. My point is that maybe instead of complicating encryption implementation with each version it is worth to make some "dump-mode" like [Kilo](https://kilo.squat.ai) or [KubeSpan](https://www.talos.dev/v1.5/talos-guides/network/kubespan/)?
What about "CIDR based strict mode" which works based on podCIDR of each node, so there could be no traffic for "newly created pod not yet in cache"?
I understand there is always complications like crossed pod and nodes subnets etc. But "dump mode" should not fix all cases, it could be just default which solves 80% of cases with same limitations as in KubeSpan.

**Comment by 3u13r** [comment]
Hey, thanks a lot for engaging in the discussion and bringing up those points. I always enjoy it when people care about their security. In general, I think that it's great that you've found a solution which fits your needs. We use Cilium as part of our [own confidential Kubernetes distro](https://github.com/edgelesssys/constellation). Therefore we need to support a multitude of deployment options while keeping the ease of use and monitoring capabilities of Cilium. 

I have some follow-up questions:
> I don't believe this is scalable solution for 200+ node cluster,

Do you have encountered any issues? One of the features this allows is that not every deployment uses full CIDR blocks for each node e.g. when Azure with direct routing is used. There the IPAM controller just assign whatever IP is free during the start of a new pod. On the other hand, one could conditionally optimize this, when PodCIDRs are used. Once scalability issues arise, I think this is worth implementing. 

> and it breaks in ~10min as per-docs, probably due to unable to renew encryption key from master

What does break? As documented the current strict mode only works on Pods. The encryption keys should be distributed by the cilium-agent making requests to the KubeAPI. So this is node-to-node communication which does not fall under the strict mode. I am happy do have a look at a full issue description to figure out what and why this breaks.

> What about "CIDR based strict mode" which works based on podCIDR of each node, so there could be no traffic for "newly created pod not yet in cache"?

Do you refer back to your point about the scalability of having an entry for each pod in the WireGuard? The strict mode itself just takes the pod CIDR of the whole pod network.

Personally I think that Cilium's encryption must also consider and secure the Cluster when e.g. Azure direct routing is used. Therefore, it's encryption (as many of Cilium's subsystems) is based on singular identities. So _I_ don't really see a way to change or "dump down" the current implementation for a subset of architectures, but maybe the actual maintainers do?

**Comment by 3u13r** [comment]
FYI: You can have a look on our current downstream implementation here: https://github.com/cilium/cilium/pull/28612.
Maybe we use it as a basis for a upstream solution.

**Comment by gadiener** [comment]
I believe this is still relevant

**Comment by jschwinger233** [comment]
Correct me if I'm wrong, strict mode CIDR doesn't care if it's pod-to-pod or node-to-node, this seems to me that we want to specify multiple strict mode CIDRs, such as `encryption.strictMode.cidrs=$pod_cidr,$node_cidr`, correct? Other than that, it's a problem of "allowlist" to always allow plain-text traffic from host to control plane.

**Comment by 3u13r** [comment]
> strict mode CIDR doesn't care if it's pod-to-pod or node-to-node

Yes, one can implement it that way. though in our downstream implementation we made a distinction to only have an port allow list that only pokes a hole into the node CIDR.

 > it's a problem of "allowlist" to always allow plain-text traffic from host to control plane. 

Essentially yes, we had to allow the etcd ports for control-plane to control-plane traffic. In our specific setup allowing KubeAPI wasn't needed as it is routed over an poxing LB outside of the strict ranges. But for a general setup this is likely needed.
